experimental/lib/Transforms/Speculation/HandshakeSpeculation.cpp

```diff
@@ -71,6 +82,193 @@ LogicalResult HandshakeSpeculationPass::placeUnits(Value ctrlSignal) {
+      for (Operation *succOp : branchOp->getResult(i).getUsers()) {
+        Value ctrl = branchCond->getResult(i);
+        routeCommitControl(markedPath, ctrl, succOp);
+      }
+    }
+
+  } else {
+    // Continue Traversal
+    for (Operation *succOp : currOp->getUsers()) {
+      routeCommitControl(markedPath, ctrlSignal, succOp);
+    }
+  }
+}
+
+LogicalResult HandshakeSpeculationPass::prepareAndPlaceCommits() {
+  // Place commits and connect to the Speculator Commit Control Signal
+  Value commitCtrl = specOp.getCommitCtrl();
+  if (failed(placeUnits<handshake::SpecCommitOp>(commitCtrl)))
+    return failure();
+
+  // Create visited set
+  llvm::DenseSet<Operation *> visited;
+  llvm::DenseSet<Operation *> markedPath;
+  visited.insert(specOp);
+
+  // Start traversal to mark the path to commits the speculator output
+  for (Operation *succOp : specOp.getDataOut().getUsers())
+    markPathToCommits(visited, markedPath, succOp);
+
+  // Follow the marked path and replicate branches
+  for (Operation *succOp : specOp.getDataOut().getUsers())
+    routeCommitControl(markedPath, commitCtrl, succOp);
+
+  return success();
+}
+
+static handshake::ConditionalBranchOp
+findControlBranch(handshake::SpeculatorOp specOp, unsigned bb) {
+  handshake::FuncOp funcOp = specOp->getParentOfType<handshake::FuncOp>();
+  assert(funcOp && "op should have parent function");
+  auto handshakeBlocks = getLogicBBs(funcOp);
+
+  for (auto condBrOp : funcOp.getOps<handshake::ConditionalBranchOp>()) {
+    if (auto brBB = getLogicBB(condBrOp); !brBB || brBB != bb)
+      continue;
+
+    for (Value result : condBrOp->getResults()) {
+      for (Operation *user : result.getUsers()) {
+
+        if (isBackedge(result, user))
+          return condBrOp;
+      }
+    }
+  }
+  funcOp->emitError() << "Could not find backedge within BB " << bb << "\n";
+  return nullptr;
+}
+
+LogicalResult HandshakeSpeculationPass::prepareAndPlaceSaveCommits() {
+  MLIRContext *ctx = &getContext();
+  OpBuilder builder(ctx);
+
+  // Don't do anything if there are no SaveCommits to place
+  if (placements.getPlacements<handshake::SpecSaveCommitOp>().empty())
+    return success();
+
+  // The save commits are a result of a control branch being in the BB
+  // The control path for the SC needs to replicate the branch
+  unsigned bb = getLogicBB(specOp).value();
+  ConditionalBranchOp controlBranch = findControlBranch(specOp, bb);
+  if (controlBranch == nullptr)
+    return failure();
+
+  // To connect a Save-Commit, two control signals are sent from the Speculator
+  // and are merged before reaching the Save-Commit.
+  // The tokens take differents paths. One needs to always reach the SC,
+  // the other should follow the actual branches similarly to the Commits
+  builder.setInsertionPointAfterValue(specOp.getSCCommitCtrl());
+  auto branchDiscardNonSpec = builder.create<handshake::SpeculatingBranchOp>(
+      controlBranch.getLoc(), controlBranch.getTrueResult() /* spec tag */,
```

Comment:
The fact that controlBranch.getTrueResult() is not safe to use for the speculative tag still has not been addressed, please change to the speculator data output ASAP.

Reply:
The specTag is now coming from the Speculator.